data/src/history/manager.rs

```diff
@@ -439,13 +439,53 @@ impl Data {
 
         let filtered = messages
             .iter()
-            .filter(|message| {
+            .enumerate()
+            .filter(|(message_index, message)| {
                 if let message::Source::Server(Some(source)) = message.target.source() {
-                    !server_messages.get(source).exclude
+                    match server_messages.get(source).exclude {
+                        Exclude::All => false,
+                        Exclude::None => true,
+                        Exclude::Smart(seconds) => 'search_previous_messages: {
```

Comment:
We've got quite a lot of nesting here, can we refactor this smart check to a function?